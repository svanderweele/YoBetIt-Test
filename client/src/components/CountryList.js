
import React from 'react';
import CountryItem from './CountryItem'
import axios from 'axios';
import Table from 'react-bootstrap/Table';
import FormControl from 'react-bootstrap/FormControl';
import InputGroup from 'react-bootstrap/InputGroup';
import Form from 'react-bootstrap/Form';
import Swal from 'sweetalert2'


class CountryList extends React.Component {


    constructor(props) {
        super(props)
        this.state = {
            countries: [],
            filter: {
                exactMatch: false,
                filterName: "Malta, Swe",
            }
        };

        this.onFilterNameChange = this.onFilterNameChange.bind(this);
        this.onFilterExactMatchChange = this.onFilterExactMatchChange.bind(this);
        this.refreshList = this.refreshList.bind(this);
    }

    componentDidMount() {
        this.refreshList();
    }

    onFilterNameChange(event) {
        this.setState({ filter: { ...this.state.filter, filterName: event.target.value } }, this.refreshList);
        this.refreshList();
    }

    onFilterExactMatchChange(event) {
        this.setState({ filter: { ...this.state.filter, exactMatch: event.target.checked } }, this.refreshList);
    }

    refreshList() {

        let url = 'http://localhost:5000/api/countries?';

        if (this.state.filter.filterName) {
            url += `country_names=${this.state.filter.filterName}`;
            console.log(url)
        }

        if (this.state.filter.exactMatch) {
            url += `&exactMatch=${this.state.filter.exactMatch}`;
        }

        axios.get(`${url}&fields=name;flag;`, { method: 'GET' })
            .then((res) => {
                this.setState({ countries: res.data.data.countries });
            })
            .catch((error) => {
                console.log(error);
            });

    }


    render() {

        return (
            <div className="container">
                <h1>Country List</h1>
                {/* Filter */}
                <InputGroup className="my-3">
                    <InputGroup.Prepend>
                        <InputGroup.Text id="basic-addon1">Filter by Name</InputGroup.Text>
                    </InputGroup.Prepend>
                    <FormControl
                        onChange={this.onFilterNameChange}
                        value={this.state.filter.filterName}
                        placeholder="Malta,Sweden"
                        aria-label="Username"
                        aria-describedby="basic-addon1"
                    />

                </InputGroup>

                <Form.Group controlId="exactMatchCheckbox">
                    <Form.Check type="checkbox" label="Exact Match?" value={this.state.filter.exactMatch} onChange={this.onFilterExactMatchChange} />
                </Form.Group>



                <Table striped bordered hover>
                    <thead>
                        <tr>
                            <th>#</th>
                            <th>Country Name</th>
                            <th>Country Flag</th>
                            <th>Active Covid Cases</th>
                        </tr>
                    </thead>
                    <tbody>
                        {this.state.countries.map(country => {
                            return <CountryItem key={country.id} country={country} />
                        })}
                    </tbody>
                </Table>
            </div>
        );

    }
}

export default CountryList;