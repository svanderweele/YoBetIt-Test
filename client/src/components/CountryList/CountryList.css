.country-flag {
  width: 100px;
}
