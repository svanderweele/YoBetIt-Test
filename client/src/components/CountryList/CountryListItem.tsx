import React from "react";
import { Country } from "../../models/Country";
import numeral from 'numeral';

interface CountryListItemProps {
  country: Country;
}

const CountryListItem = (props: CountryListItemProps) => {
  const tdStyle = {
    verticalAlign: "middle",
  };

  return (
    <tr>
      <td style={tdStyle}>{props.country.id}</td>
      <td style={tdStyle}>{props.country.name}</td>
      <td style={tdStyle}>
        <img className='img-fluid' width='100px' src={props.country.flag} alt="Imagine a flag was here" />
      </td>
      <td style={tdStyle}>{numeral(props.country.activeCoronaCases).format('0,0')}</td>
    </tr>
  );
};

export default CountryListItem;
