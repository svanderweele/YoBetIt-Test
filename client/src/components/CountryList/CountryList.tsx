import React from "react";
import { Table, Card, Container } from "react-bootstrap";
import "./CountryList.css";
import {Country, GetCountriesResponse} from "../../models/Country";
import CountryItem from "../CountryItem/CountryItem";
import FilterName from "../FilterName/FilterName";

const CountryList = () => {
  const [filterState, setFilterState] = React.useState({
    name: "",
    isExactNameMatch: false,
  });

  const[countries, setCountries] = React.useState<Country[] | null>(null);

  const retrieveCountries = () => {

    let urlQueryParams = [];

    if(filterState.isExactNameMatch){
      urlQueryParams.push('isExactNameMatch=true');
    }

    if(filterState.name){
      urlQueryParams.push(`countryNames=${filterState.name}`);
    }

    fetch(`api/countries?${urlQueryParams.map((query,index) => index > 0 ? `&${query}` : query).join('')}`)
      .then((res) => res.json())
      .then((response : GetCountriesResponse) => {
        setCountries(response.countries);
      });
  };

  React.useEffect(() => {
    retrieveCountries();
  }, [filterState]);

  const onFilterNameChange = (newName: string) => {
    setFilterState({ ...filterState, name: newName });
  };

  const onFilterExactMatchChange = (value: boolean) => {
    setFilterState({ ...filterState, isExactNameMatch: value });
  };

  return (
    <Container className="mt-2">
      <Card>
        <Card.Body>
          <Card.Title>Country List</Card.Title>
          <FilterName
            onFilterNameChanged={onFilterNameChange}
            onFilterExactMatch={onFilterExactMatchChange}
          />
          <Table bordered>
            <thead>
              <tr>
                <th>Flag</th>
                <th>Name</th>
                <th>Covid-19 Active Cases</th>
              </tr>
            </thead>
            <tbody>
              {countries?.map((country) => {
                return <CountryItem key={country.name} country={country} />;
              })}
            </tbody>
          </Table>
        </Card.Body>
      </Card>
    </Container>
  );
};

export default CountryList;
