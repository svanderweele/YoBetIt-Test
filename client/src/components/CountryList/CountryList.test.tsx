import React from "react";
import { shallow } from "enzyme";
import CountryList from "./CountryList";
describe("CountryList", () => {
  it('should render correctly', () => {
    const countryList = shallow(<CountryList />);
    expect(countryList).toMatchSnapshot();
  });
});
