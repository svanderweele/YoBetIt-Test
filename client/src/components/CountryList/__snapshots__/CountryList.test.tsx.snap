// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryList should render correctly 1`] = `
<Card
  body={false}
>
  <CardBody>
    <CardTitle>
      Country List
    </CardTitle>
    <CountryListFilter
      onFilterExactMatch={[Function]}
      onFilterNameChanged={[Function]}
    />
    <ForwardRef
      bordered={true}
    >
      <thead>
        <tr>
          <th>
            Id
          </th>
          <th>
            Name
          </th>
          <th>
            Flag
          </th>
          <th>
            Active Corona Cases
          </th>
        </tr>
      </thead>
      <tbody />
    </ForwardRef>
  </CardBody>
</Card>
`;
